src/wovnio/html/HtmlConverter.php

```diff
@@ -276,10 +276,31 @@ private function insertHreflangTags($html)
         return $this->insertAfterTag($parent_tags, $html, implode('', $hreflangTags));
     }
 
+    private function insertCanonicalTag($html)
+    {
+        if ($this->isNoindexLang($this->headers->requestLang())) {
+            return $html;
+        }
+
+        $canonical_tag = '<link rel="canonical" href="' . $this->buildCanonicalUrl($this->headers->requestLang()) . '">';
```

Comment:
`buildCanonicalUrl` is simply delegating to Headers again. Let Headers provide the translated canonical url to begin with. `getCanonicalUrl()` can just return the translated one.